File: cv_nextjs/src/helpers/weightTechs.ts
import { Job } from "../models/classes/Job"
import { Project } from "../models/classes/Project"

export const techsWithMonthsDuration = (
        items: (Job | Project)[]
    ) : Map<number, number> => {

    const weights = new Map()

    items.forEach((jobOrperiod) => {
 
        const techs = jobOrperiod.tech

        techs.forEach((tech) => {
            if (!weights.has(tech.uid)) {
                weights.set(tech.uid, 0)
            }

            const curr = weights.get(tech.uid)

            weights.set(tech.uid, curr + 1)
        })
    })

    return weights
}

export const techWeight = (
        techId: number, 
        techWeights: Map<number, number>, 
        maxTechWeight: number
    ) : number => {

    let thisWeight = techWeights.get(techId)

    if (!thisWeight) {
        return 0
    }

    let percent = (100  * thisWeight) / maxTechWeight;

    if (percent > 80) {
        return 24
    } else if (percent > 60) {
        return 22
    } else if (percent > 40) {
        return 19
    } else if (percent > 20) {
        return 14
    } else {
        return 12
    }
}

File: cv_rust_wasm/src/util/weight_techs.rs
use std::collections::HashMap;

use log::info;

use crate::models::{
    job_model::JobModel, 
    project_model::ProjectModel,
};

pub fn techs_with_months_duration (jobs: Vec<&JobModel>, projects: Vec<&ProjectModel>) -> HashMap<usize, i32> {

    let mut weights: HashMap<usize, i32> = HashMap::new();

    for job in jobs {

        // let job_duration = job.period.months_duration();

        for tech in &job.tech {

            if !weights.contains_key(&tech.uid) {
                weights.insert(tech.uid, 0);
            }

            weights.entry(tech.uid).and_modify(|x| *x += 1);
        }
    }

    for project in projects {

        // let project_duration = project.period.months_duration();

        for tech in &project.tech {

            if !weights.contains_key(&tech.uid) {
                weights.insert(tech.uid, 0);
            }

            weights.entry(tech.uid).and_modify(|x| *x += 1);
        }
    }

    weights
}

pub fn tech_weight (tech_id: usize, tech_weights: &HashMap<usize, i32>, max_tech: i32) -> i32 {

    let this_weight = tech_weights.get(&tech_id);

    if this_weight.is_none() {
        
        return 0
    }

    let percent = (100 *  this_weight.unwrap()) / max_tech;

    if percent > 75 {
        return 26
    } else if percent > 50 {
        return 22
    } else if percent > 25 {
        return 18
    } else {
        return 14
    }
}